### DISCUSSION TITLE
Skia renderer with Linux backend

### DISCUSSION THREAD

**Discussion by igiona** [slint-ui/slint#6611] [Q&A] [answered]
I'm trying to cross-compile a Slint application using the Linux backend and the Skia renderer.
Unfortunately, I get several errors.
The first

```bash
error: failed to run custom build command for `skia-bindings v0.78.2`
[...]
  [8/1528] compile ../../../../../../../../../../../../home/igiona/.cargo/registry/src/index.crates.io-6f17d22bba15001f/skia-bindings-0.78.2/skia/src/ports/SkFontConfigInterface.cpp
  FAILED: obj/src/ports/fontmgr_FontConfigInterface.SkFontConfigInterface.o
  clang++ -MD -MF obj/src/ports/fontmgr_FontConfigInterface.SkFontConfigInterface.o.d -DNDEBUG -DSK_CODEC_DECODES_BMP -DSK_CODEC_DECODES_WBMP -DSK_R32_SHIFT=16 -DSK_ENABLE_PRECOMPILE -DSK_GANESH -DSK_DISABLE_TRACING -DSK_USE_PERFETTO -DSK_GAMMA_APPLY_TO_A8 -DSK_ENABLE_AVX512_OPTS -DSKIA_IMPLEMENTATION=1 -DSK_TYPEFACE_FACTORY_FREETYPE -I../../../../../../../../../../../../home/igiona/.cargo/registry/src/index.crates.io-6f17d22bba15001f/skia-bindings-0.78.2/skia -Wno-attributes -ffp-contract=off -fstrict-aliasing -fPIC -fvisibility=hidden -march=armv7-a -mfpu=neon -mthumb -O3 -fdata-sections -ffunction-sections -I=/usr/include/freetype2 -I/usr/include/freetype2 -O3 -I/usr/armv7-linux-gnueabihf/include/c++/7/armv7-linux-gnueabihf -I/usr/armv7-linux-gnueabihf/include --target=armv7-unknown-linux-gnueabihf -std=c++17 -fvisibility-inlines-hidden -fno-exceptions -fno-rtti -c ../../../../../../../../../../../../home/igiona/.cargo/registry/src/index.crates.io-6f17d22bba15001f/skia-bindings-0.78.2/skia/src/ports/SkFontConfigInterface.cpp -o obj/src/ports/fontmgr_FontConfigInterface.SkFontConfigInterface.o
  In file included from ../../../../../../../../../../../../home/igiona/.cargo/registry/src/index.crates.io-6f17d22bba15001f/skia-bindings-0.78.2/skia/src/ports/SkFontConfigInterface.cpp:7:
  In file included from ../../../../../../../../../../../../home/igiona/.cargo/registry/src/index.crates.io-6f17d22bba15001f/skia-bindings-0.78.2/skia/include/ports/SkFontConfigInterface.h:11:
  In file included from ../../../../../../../../../../../../home/igiona/.cargo/registry/src/index.crates.io-6f17d22bba15001f/skia-bindings-0.78.2/skia/include/core/SkFontStyle.h:11:
  In file included from ../../../../../../../../../../../../home/igiona/.cargo/registry/src/index.crates.io-6f17d22bba15001f/skia-bindings-0.78.2/skia/include/core/SkTypes.h:21:
  In file included from ../../../../../../../../../../../../home/igiona/.cargo/registry/src/index.crates.io-6f17d22bba15001f/skia-bindings-0.78.2/skia/include/private/base/SkAssert.h:15:
  /usr/lib/gcc-cross/arm-linux-gnueabihf/11/../../../../include/c++/11/cstddef:49:10: fatal error: 'bits/c++config.h' file not found
     49 | #include <bits/c++config.h>
        |          ^~~~~~~~~~~~~~~~~~
  1 error generated.

[...]
``` 

This seems to me related to some kind of issues with sysroot or so. 
I did not manage to pass sysroot to the C compiler though, or at least the error never changed.
I therefore tried to source the cross-compile environment file that Yocto generates along with the SDK, but then I stumble across another error:

```bash
error: failed to run custom build command for `skia-bindings v0.78.2`

[...]

  [1/1521] compile ../../../../../../../../../../../../home/igiona/.cargo/registry/src/index.crates.io-6f17d22bba15001f/skia-bindings-0.78.2/skia/src/ports/SkFontConfigInterface.cpp
  FAILED: obj/src/ports/fontmgr_FontConfigInterface.SkFontConfigInterface.o
  arm-org-linux-gnueabi-g++  -march=armv7ve -mthumb -mfpu=neon -mfloat-abi=hard -fstack-protector-strong  -O2 -D_FORTIFY_SOURCE=2 -Wformat -Wformat-security -Werror=format-security --sysroot=/home/igiona/sdks/lcd_sdk2/sysroots/armv7vet2hf-neon-org-linux-gnueabi -MD -MF obj/src/ports/fontmgr_FontConfigInterface.SkFontConfigInterface.o.d -DNDEBUG -DSK_CODEC_DECODES_BMP -DSK_CODEC_DECODES_WBMP -DSK_R32_SHIFT=16 -DSK_ENABLE_PRECOMPILE -DSK_GANESH -DSK_DISABLE_TRACING -DSK_USE_PERFETTO -DSK_GAMMA_APPLY_TO_A8 -DSK_ENABLE_AVX512_OPTS -DSKIA_IMPLEMENTATION=1 -DSK_TYPEFACE_FACTORY_FREETYPE -I../../../../../../../../../../../../home/igiona/.cargo/registry/src/index.crates.io-6f17d22bba15001f/skia-bindings-0.78.2/skia -Wno-attributes -fstrict-aliasing -fPIC -fvisibility=hidden -march=armv7-a -mfpu=neon -mthumb -O3 -fdata-sections -ffunction-sections -I=/usr/include/freetype2 -I/usr/include/freetype2 --sysroot=/home/igiona/sdks/lcd_sdk2/sysroots/armv7vet2hf-neon-org-linux-gnueabi -O3 -I/usr/armv7-linux-gnueabihf/include/c++/7/armv7-linux-gnueabihf -I/usr/armv7-linux-gnueabihf/include --target=armv7-unknown-linux-gnueabihf -std=c++17 -fvisibility-inlines-hidden -fno-exceptions -fno-rtti -c ../../../../../../../../../../../../home/igiona/.cargo/registry/src/index.crates.io-6f17d22bba15001f/skia-bindings-0.78.2/skia/src/ports/SkFontConfigInterface.cpp -o obj/src/ports/fontmgr_FontConfigInterface.SkFontConfigInterface.o
  arm-org-linux-gnueabi-g++: error: unrecognized command-line option '--target=armv7-unknown-linux-gnueabihf'

[...]
```

Any idea what could be wrong here?

**Comment by ogoffart** [comment]
There is some documentation of some env variables to set here: https://github.com/rust-skia/rust-skia/tree/master/skia-bindings#cross-compiling-for-linux

**Reply by tronical** [reply]
For the first error, I can see indeed that the sysroot is missing. The link that Olivier pasted is indeed the right starting point probably, it might be sufficient for you to set `SDKTARGETSYSROOT`.

The `unrecognized command-line option '--target=armv7-unknown-linux-gnueabihf'` suggests that there's a mismatch between the triplet your Yocto uses and what rust knows (which is consequently turned into a `--target=` option). In your Yocto SDK environment, do you have the `CLANGCXX` and `CLANGCC` environment variables set? If yes, what is their value? They should have a `--target=...` option in them, and the Skia build should "extract" and re-use them.

**Reply by igiona** [reply]
@ogoffart thank for the link!

@tronical No, the Yocto SDLK environment doesn't have neither of bot `CLANGCXX` / `CLANGCC` .
There are `CC` and `CXX` though, with all arch params etc.

The triplet created by Yocto is indeed different than rustc/cargo.
I build with `cargo build --release --target armv7-unknown-linux-gnueabihf` and our triplet is `arm-org-linux-gnueabi` .

I set `HOST_CC` as suggested in the link above, with the same result.

As far as I know, CC doesn't have `--target` option??

**Reply by tronical** [reply, marked as answer]
Ah, I think I realize what you might be missing.

The Skia build requires clang, and that clang needs to be entirely aware of your yocto target config and gcc library paths. Sometimes it works to use the host clang, but it's better to use meta-clang. Try this:

1. Add [`meta-clang`](https://github.com/kraj/meta-clang) to your layers.
2. Add clang to your generated SDK like this: https://github.com/kraj/meta-clang?tab=readme-ov-file#adding-clang-in-generated-sdk-toolchain

From the environment of the resulting SDK, you'll have `CLANGCC` and `CLANGCXX` set, the Skia build will pick them up automatically, and it ... should compile.

**Reply by igiona** [reply]
Fantastic @tronical ! That was it 🥳 !

Now I've got the binary, but I'm still not able to run the application on the device :'(
When I try to execute the application I get:

```
root@07083577:/usr/bin/ui$ ls -l
-rwxr-xr-x    1 root     root      21472636 Jan  1 01:55 ui-rs
root@07083577:/usr/bin/ui$ ./ui-rs
-sh: ./ui-rs: not found
```

Do I have to pass some command line parameters?

**Reply by tronical** [reply]
Could you try running it like this?

`LD_DEBUG=all ./ui-rs`

Also, what does `file ./ui-rs` say?

**Reply by igiona** [reply]
I don't have `file` on the target, but on the host I get:
```
ui-rs: ELF 32-bit LSB pie executable, ARM, EABI5 version 1 (GNU/Linux), dynamically linked, interpreter /lib/ld-linux.so.3, BuildID[sha1]=cb9b20138a725b9b473cfd85f6b5c7bfbbf5a6bc, for GNU/Linux 3.2.0, not stripped
```

Adding the debug variable doesn't help, I get the same result.

```
LD_DEBUG=all ui-rs
-sh: ui-rs: not found
```

**Reply by tronical** [reply]
Sorry, I meant `LD_DEBUG=all ./ui-rs` or `LD_DEBUG=all /usr/bin/ui/ui-rs`.

**Reply by tronical** [reply]
Or could it be that something went wrong during the copy of the file to the device?

**Reply by igiona** [reply]
Okay, I found the issue, this sounded fishy `... dynamically linked, interpreter /lib/ld-linux.so.3 ...` ?
I don't have this library on my target machine, but I saw that another rust application build for the same target) was slightly different:
``` 
ELF 32-bit LSB pie executable, ARM, EABI5 version 1 (SYSV), dynamically linked, interpreter /lib/ld-linux-armhf.so.3, for GNU/Linux 3.2.0, not stripped
```

This solved (workaround) the issue:
```
ln -s /lib/ld-linux-armhf.so.3 /lib/ld-linux.so.3
```

I wonder how can be fixed properly!?

Now my Slint app runs on the target machine ❤️ 
Thank you for the support 🙏

**Reply by tronical** [reply]
> Now my Slint app runs on the target machine 

Yay :)

> I wonder how can be fixed properly!?

This sounds like a mismatch between your Yocto Build environment and toolchain and the image that you've flashed. I suppose other binaries built by the same environment also have `/lib/ld-linux-armhf.so.3` as their loader configured?

**Reply by igiona** [reply]
>This sounds like a mismatch between your Yocto Build environment and toolchain and the image that you've flashed. I suppose other binaries built by the same environment also have /lib/ld-linux-armhf.so.3 as their loader configured?

That can be, but when I cross-compile a purely rust app, I just need a compatible linker (with the same glibc version), I usually use a standard gcc (gcc-arm-10.x) and it works fine.

I'll see what happens when the Slint app is built within Yocto, maybe it's going to "just work".
